### PR TITLE
feat(#15): Add support for JUnit5

### PR DESCRIPTION
**Pull request by lordofthejars**

### CODE REVIEW

**Comment by bartoszmajsak** [comment]
I'm reviewing this code, but since it's `Status: Work In Progress` could you make a checklist of what still needs to be done? Thanks!

File: junit-test-result-parser/pom.xml [hunk 1]
```diff
@@ -18,6 +18,7 @@
     <dependency>
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
+      <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (lordofthejars) [reply junit-test-result-parser/pom.xml]: No it is not need, but in my way to try to find which artifact is adding junit4 as compile I decided to force everywhere. I am going to delete.

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply junit-test-result-parser/pom.xml]: That's what I thought :)

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        return getJunitDepVersion() != null && isAnyJunit5() && junit5SurefirePlatformVersion != null;
+    }
+
+    public String getDepCoordinates() {
+        return "org.junit.platform:junit-platform-surefire-provider:" + junit5SurefirePlatformVersion;
+    }
+
+    private String retrieveJunit5SurefirePlatformVersion() {
+
+        final Path junit5PlatformVersionFile = new LocalStorage(Paths.get("").toFile()).duringExecution()
```
[the remark this answers is not in the document]
Comment (lordofthejars) [reply surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java]: Already discussed in bjn

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnit5Test.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    private final String junitVersion;
+
+    public LoaderVersionExtractorJUnit5Test(String junitVersion) {
+        this.junitVersion = junitVersion;
+    }
+
+    @Parameterized.Parameters
+    public static Collection<String> data() {
+        return Arrays.asList("5.0.0");
```
[the remark this answers is not in the document]
Comment (lordofthejars) [reply surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnit5Test.java]: Already discussed in bjn

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    private final String surefirejunitVersion;
+
+    public LoaderVersionExtractorSurefireJUnit5Test(String junitSurefireVersion) {
+        this.surefirejunitVersion = junitSurefireVersion;
+    }
+
+    @Parameterized.Parameters
+    public static Collection<String> data() {
+        return Arrays.asList("1.0.0");
```
[the remark this answers is not in the document]
Comment (lordofthejars) [reply surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java]: Already discussed in bjn

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java [hunk 5]
```diff
@@ -108,6 +108,23 @@ public void should_fail_when_autocorrect_repeats_the_strategy()
...
     }
 
+    @Test
+    public void should_remove_junit5_platform_dependency() {
+        // given
+        Model model = prepareModelWithSurefirePlugin("2.19.1");
+        final Configuration conf = configureWithAutocorrect("nwe");
+        final DependencyResolver dependencyResolver = new DependencyResolver(conf);
+        final Plugin plugin = model.getBuild().getPlugins().get(0);
+        plugin.addDependency(new DependencyResolver.JUnit5SurefireProviderDependency());
```
[the remark this answers is not in the document]
Comment (lordofthejars) [reply mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: I understand your point but, it is not so easy because in some tests you need both `model` and `dependencyResolver`. One possible solution would be create a method with so many parameters so all possibilities can be fit with this method call.

The other option would be create some classes using builder pattern, but my question is is it worth for just 4 lines when the code is duplicated in just 3 test methods to do this?

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java [hunk 6]
```diff
@@ -108,6 +108,23 @@ public void should_fail_when_autocorrect_repeats_the_strategy()
 
     }
 
+    @Test
+    public void should_remove_junit5_platform_dependency() {
+        // given
+        Model model = prepareModelWithSurefirePlugin("2.19.1");
+        final Configuration conf = configureWithAutocorrect("nwe");
```
[the remark this answers is not in the document]
Comment (lordofthejars) [reply mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: fixed.

File: functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/junit5/LocalChangesTestsSelectionExecutionFunctionalTest.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import static org.arquillian.smart.testing.ftest.testbed.configuration.Mode.SELECTING;
+import static org.arquillian.smart.testing.ftest.testbed.configuration.Strategy.CHANGED;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class LocalChangesTestsSelectionExecutionFunctionalTest {
+
+    @ClassRule
+    public static final GitClone GIT_CLONE = new GitClone("https://github.com/lordofthejars/junit5-demo.git");
+
```
Comment (dipak-pawar) [inline functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/junit5/LocalChangesTestsSelectionExecutionFunctionalTest.java]: Are we going to use the same repository or we are going to transfer ownership to arquillian org?

Comment (lordofthejars) [reply functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/junit5/LocalChangesTestsSelectionExecutionFunctionalTest.java]: Yes, I am not sure also how we should named in arquillian, because it is not really a dogfooding project. @bartoszmajsak WDYT about this?

My point now is that this is a really small project just to check that the file creation and read is done correctly, which is enough in case of JUnit5. So maybe call it `smart-testing-junit5-demo` in arquillian organization is enough.

Comment (bartoszmajsak) [reply functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/junit5/LocalChangesTestsSelectionExecutionFunctionalTest.java]: Yes, this is rather a demo project which we can transfer to the arq organization. We only do thin smoke test to be sure integration is fine, as the core logic which are tests are focusing on are not the test execution (so junit4 or junit5), but selection logic so our provider which delegates to aforementioned runners.

Let's move it as @lordofthejars suggests.

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java [hunk 8]
```diff
@@ -39,8 +44,25 @@ void configureTestRunner(Model model) {
 
             dependencyResolver.addRequiredDependencies(model);
 
+
+            final LocalStorage localStorage = new LocalStorage(Paths.get("").toFile());
```
Comment (MatousJobanek) [inline mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: location

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java [hunk 9]
```diff
@@ -39,8 +44,25 @@ void configureTestRunner(Model model) {
...
-                .forEach(dependencyResolver::addAsPluginDependency);
+                .forEach(plugin -> {
+                    dependencyResolver.addAsPluginDependency(plugin);
+
+                    final Optional<Dependency> dependency = dependencyResolver.findJUnit5PlatformDependency(plugin);
+                    dependency.ifPresent(d -> {
+                        try {
+                            localStorage.duringExecution()
+                                .temporary()
+                                .file("junit5PlatformVersion")
```
Comment (MatousJobanek) [inline mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: constant

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java [hunk 10]
```diff
@@ -39,8 +44,25 @@ void configureTestRunner(Model model) {
 
             dependencyResolver.addRequiredDependencies(model);
 
+
+            final LocalStorage localStorage = new LocalStorage(Paths.get("").toFile());
             effectiveTestRunnerPluginConfigurations
-                .forEach(dependencyResolver::addAsPluginDependency);
+                .forEach(plugin -> {
+                    dependencyResolver.addAsPluginDependency(plugin);
```
Comment (MatousJobanek) [inline mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: what if there are different versions for each plugin

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java [hunk 11]
```diff
@@ -75,7 +76,7 @@ public void should_autocorrect_strategy()
         Model model = prepareModelWithSurefirePlugin("2.20");
         final Configuration conf = configureWithAutocorrect("nwe");
         final DependencyResolver dependencyResolver = new DependencyResolver(conf);
-        model.addDependency(new DependencyResolver.SurefireApiDependency("2.19.1"));
+        model.addDependency(new DependencyResolver.SurefireApiDependency("2.20"));
```
Comment (MatousJobanek) [inline mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: why this change?

File: strategies/affected/pom.xml [hunk 12]
```diff
@@ -32,7 +32,6 @@
     <dependency>
       <groupId>junit</groupId>
       <artifactId>junit</artifactId>
-      <scope>test</scope>
```
Comment (MatousJobanek) [inline strategies/affected/pom.xml]: why is the scope removed?

Comment (lordofthejars) [reply strategies/affected/pom.xml]: Because it is inherited

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java (hunk 11 above)
Comment (lordofthejars) [reply mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: Well it is a mocked instance, and you can change it without any worries, it works with any number there.

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java (hunk 10 above)
Comment (lordofthejars) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: what do you mean about different versions of each plugin here, if I understand you correctly you mean what's happening if you have different versions of junit5 in the same project, which I think it is something really strange if this is what I think.

Comment (lordofthejars) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: Also notice that this plugin is added from a foreach which is all plugins `effectiveTestRunnerPlugin`

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java (hunk 11 above)
Comment (MatousJobanek) [reply mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: so, again, why you have changed it? is there any reason for it?

File: strategies/affected/pom.xml (hunk 12 above)
Comment (MatousJobanek) [reply strategies/affected/pom.xml]: that you can say about the rest of the test dependencies... I would prefer to have everything specified there so I don't have to go to the parent pom checking what is specified there...

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java]: that should be taken from this method: https://github.com/arquillian/smart-testing/blob/master/surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/SmartTestingSurefireProvider.java#L88

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java (hunk 11 above)
Comment (lordofthejars) [reply mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: Well because the model is prepared with 2.20, so to make it more real the dependency should be 2.20 as well isn't it?See line 76

File: strategies/affected/pom.xml (hunk 12 above)
Comment (lordofthejars) [reply strategies/affected/pom.xml]: Originally there were not there the scope. Then I added for testing one thing but I committed by error, and then @bartoszmajsak (And I agree with him) asked me to remove again.

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractor.java [hunk 13]
```diff
@@ -42,6 +47,14 @@ public static String getJunitVersion() {
         return getVersionFromClassLoader(LIBRARY_JUNIT, currentThread().getContextClassLoader());
     }
 
+    public static String getJUnit5Version() {
+        return getVersionFromClassLoader(LIBRARY_JUNIT_5, currentThread().getContextClassLoader());
+    }
+
+    public static String getSurefireJUnit5Version() {
+        return getVersionFromClassLoader(LIBRARY_SUREFIRE_JUNIT_5, currentThread().getContextClassLoader());
+    }
```
Comment (MatousJobanek) [inline surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractor.java]: Do we really need these methods? To simulate the Surefire's behavior correctly, if the JUnit 5 provider is present, then we should delegate to this provider, no matter what is on the classpath...

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        super(LoaderVersionExtractor.getJUnit5Version());
+        junit5SurefirePlatformVersion = retrieveJunit5SurefirePlatformVersion();
+    }
+
+    public String getProviderClassName() {
+        return "org.junit.platform.surefire.provider.JUnitPlatformProvider";
+    }
+
+    public boolean isApplicable() {
+        return getJunitDepVersion() != null && isAnyJunit5() && junit5SurefirePlatformVersion != null;
```
Comment (MatousJobanek) [inline surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java]: yeah, here we should just check if the `junit5SurefirePlatformVersion` is not null

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnit5Test.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.Collection;
+import org.jboss.shrinkwrap.resolver.api.maven.Maven;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Parameterized.class)
+public class LoaderVersionExtractorJUnit5Test {
```
Comment (MatousJobanek) [inline surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnit5Test.java]: this class seems to be very similar to the https://github.com/arquillian/smart-testing/blob/master/surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnitTest.java

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.Collection;
+import org.jboss.shrinkwrap.resolver.api.maven.Maven;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(Parameterized.class)
+public class LoaderVersionExtractorSurefireJUnit5Test {
```
Comment (MatousJobanek) [inline surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java]: same as the previous one

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java (hunk 10 above)
Comment (MatousJobanek) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: I meant a different version of JUnit5 provider for surefire and failsafe...

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java (hunk 14 above)
Comment (lordofthejars) [reply surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java]: This is what I am doing.

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnit5Test.java (hunk 15 above)
Comment (lordofthejars) [reply surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorJUnit5Test.java]: This test probably needs to be removed.

File: mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java (hunk 11 above)
Comment (MatousJobanek) [reply mvn-extension/src/test/java/org/arquillian/smart/testing/mvn/ext/dependencies/SurefireApiDependencyTest.java]: Yeah, I missed that line :-). Thx
anyway, I'm not sure if adding the surefire dependency is really necessary here... But that can be part of another PR ;-)

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java (hunk 16 above)
Comment (lordofthejars) [reply surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java]: If you check the project before this PR you'll notice that each component we detect has its own LoaderVersion test class, one for TestNG, other for JUnit component and so on. So I am just following the same pattern as before. We can change, but in this case this was implemented following the same approach.

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractor.java (hunk 13 above)
Comment (lordofthejars) [reply surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractor.java]: In fact needs to be removed.

File: strategies/affected/pom.xml (hunk 12 above)
Comment (MatousJobanek) [reply strategies/affected/pom.xml]: originally when? looking at the state before your commit, it was there ;-) https://github.com/arquillian/smart-testing/blob/1498d4051942f3b8725712bcd5fbe547cbd2e82b/strategies/affected/pom.xml#L35

Comment (lordofthejars) [reply strategies/affected/pom.xml]: check other pom.xmls files and check that it was not present

File: surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java (hunk 14 above)
Comment (MatousJobanek) [reply surefire-provider/src/main/java/org/arquillian/smart/testing/surefire/provider/info/JUnit5ProviderInfo.java]: I mean that it should check only if the version is present:  
` return junit5SurefirePlatformVersion != null;`
see this comment: https://github.com/arquillian/smart-testing/pull/241#discussion_r147354761

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java (hunk 10 above)
Comment (lordofthejars) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: I think this is a really corner case, I mean I thought about this possibility but adding junit5 surefire platform with one version and in the same project another junit5 platform version for failsafe I would say that it is really really strange.

File: surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java (hunk 16 above)
Comment (MatousJobanek) [reply surefire-provider/src/test/java/org/arquillian/smart/testing/surefire/provider/LoaderVersionExtractorSurefireJUnit5Test.java]: Yeah, right... It just looks weird that it is the very same test with one parameter. 
But I agree.

File: strategies/affected/pom.xml (hunk 12 above)
Comment (MatousJobanek) [reply strategies/affected/pom.xml]: Then, from my point of view, it should be there. Otherwise, you can remove the scope from all other test dependencies

Comment (MatousJobanek) [reply strategies/affected/pom.xml]: We had this kind of conversation already and I do remember that we agreed on having it there, otherwise, it brings a lot of confusions...

Comment (lordofthejars) [reply strategies/affected/pom.xml]: Something to discuss on Cabal I guess to know how to proceed in general.

Comment (lordofthejars) [reply strategies/affected/pom.xml]: Yes I remember this but since I saw that test scope was not in any of the modules, then I decided to remove as well to follow the same convention. In fact if you notice the only one that it had before was `affected`

Comment (MatousJobanek) [reply strategies/affected/pom.xml]: as for the state before your commit: 
`find ./ -name "pom.xml" | xargs grep "artifactId>junit<" -A 1`
```xml
./functional-tests/test-bed/pom.xml:      <artifactId>junit</artifactId>
./functional-tests/test-bed/pom.xml-      <scope>compile</scope>
--
./functional-tests/git-rules/pom.xml:      <artifactId>junit</artifactId>
./functional-tests/git-rules/pom.xml-      <scope>compile</scope>
--
./strategies/changed/pom.xml:      <artifactId>junit</artifactId>
./strategies/changed/pom.xml-      <scope>test</scope>
--
./strategies/failed/pom.xml:      <artifactId>junit</artifactId>
./strategies/failed/pom.xml-      <scope>test</scope>
--
./strategies/affected/pom.xml:      <artifactId>junit</artifactId>
./strategies/affected/pom.xml-      <scope>test</scope>
--
./surefire-provider/pom.xml:      <artifactId>junit</artifactId>
./surefire-provider/pom.xml-      <scope>test</scope>
--
./junit-test-result-parser/pom.xml:      <artifactId>junit</artifactId>
./junit-test-result-parser/pom.xml-    </dependency>
--
./mvn-extension/pom.xml:      <artifactId>junit</artifactId>
./mvn-extension/pom.xml-    </dependency>
--
./core/pom.xml:      <artifactId>junit</artifactId>
./core/pom.xml-    </dependency>
```
so, I see more cases where the scope was specified... still, if you wanted to keep it unified, you should have removed it from all other test dependencies, which means everyting that is in the parent pom file: https://github.com/arquillian/smart-testing/blob/1498d4051942f3b8725712bcd5fbe547cbd2e82b/pom.xml#L85-L114...

File: mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java (hunk 10 above)
Comment (MatousJobanek) [reply mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/MavenProjectConfigurator.java]: There is another case - consider having surefire with JUnit 5 and failsafe without it...
